### DISCUSSION TITLE
Late Function Binding During Evaluation

### DISCUSSION THREAD

**Discussion by hyp0th3rmi4** [cel-expr/cel-go#1081] [Ideas]
This discussion is to revive a discussion in the following issue: #356, which essentially asks for the ability to binding functions at evaluation time rather than during parsing.

I have developed a solution, and I would like to discuss whether it is something that might be considered useful. The reason why I think this might be useful, is because we use google-cel in a business rules engine, which operates as a micro-service, and often we are asked to mutate per request some aspects. 

One example would be to run business rules at a given time in the future, hence the need to override (per request) what is the value of the `now()` function defined in one of the libraries we use to compile the rules. Another use case is context based logging in the functions used for the evaluation.

Based on my initial investigation, I think it is rather feasible and should not cause much trouble. The point where I got stuck was the ability to visit the interpretable so that bindings got replaced. While this is possible in the planner, the planning phase occurs at compilation time.

Any thoughts or interest in this matter?

**Comment by TristonianJones** [comment]
@hyp0th3rmi4 This is a really important feature and I expect we'll implemented it in Q1 since we have some other projects which will rely on accumulated state through a side-channel, such as expression debugging or coverage instrumentation. The feature will require that a signature for the overload is provided to the runtime somehow which will indicate that the `Activation` should be checked for the implementation.

In terms of `now`, it's probably best to implement the `Activation` interface such that you can lazily provide the right timestamp on demand.

**Reply by hyp0th3rmi4** [reply]
Thanks @TristonianJones !

**Reply by hyp0th3rmi4** [reply]
Hi @TristonianJones has there been any update on planning/roadmapping this capability?

**Reply by TristonianJones** [reply]
Portions of the implementation are ready to be integrated back into CEL, but the dispatch implementation is still pending.

**Reply by hyp0th3rmi4** [reply]
Hi @TristonianJones I have supplied a PR with a reference / prototype implementation that is fully functional and tested here:

https://github.com/google/cel-go/pull/1172

The implementaiton is based on `functions.Overload`, rather than `decls.OverloadDecl`. As a result of this at runtime we don't have the necessary information to fully verify the signatures. It is a bit of a low level approach. The changes to uplift to include information about the full type signature aren't big on that implementation.

If it is something of interest I can work on this.

**Reply by hyp0th3rmi4** [reply]
Hi @TristonianJones any update on the implementation of function binding during evaluation? 
Would you be able to run /gcbrun on #1172, if there is any interest?

**Reply by mneff-roblox** [reply]
+1

**Reply by mneff-roblox** [reply]
@TristonianJones are you still open to receiving an implementation on this?

**Reply by TristonianJones** [reply]
Hi @mneff-roblox, I'm sorry I missed seeing this. Yes, I'm open. You may also want to consider using the new async functions if those are a good fit for your use case.

**Reply by TristonianJones** [reply]
I've gone ahead and implemented this in #1483
